lib/types.nix

```
@@ -300,6 +300,19 @@ rec {
       inherit (str) merge;
     };
 
+    # Allow a newline character at the end and trim it in the merge function.
+    singleLineStr =
+      let
+        inherit (strMatching "[^\n\r]*\n?") check merge;
+      in
+      mkOptionType {
+        name = "singleLineStr";
+        description = "string that doesn't contain [\\n\\r]";
```

this isn't quite correct anymore and could be misleading, but not sure how to phrase it well without introducing odd artifacts in composite types (eg how `listOf` uses a `list of ${elem}s` as its description)

maybe something like this?
```suggestion
        description = "(optionally [\\r\\n]-terminated) single-line string";
```

---

Thanks :)

---

oh heck, our suggestion was wrong. you're only checking for a `\n` terminator, not for `[\r\n]`. didn't even notice that until now, when we wrote it we thought you checked them both. :(

---

I wondered about matching both but I saw that everywhere in nixpkgs there were splits and concats with just `\n` and very few occurrences of `\r` so I thought not allowing `\r` inside the string is relatively harmless (and there's a few occurrences of that in nixpkgs) but actually expecting a `\r\n` at the end would be a first.

---

yeah, expecting an `\n` at the end is totally fine. should've just written `newline-terminated` instead of `[\r\n]-terminated` :D

---

I amended that.